### ISSUE TITLE
Consider hiding SSL_ctrl and promoting macros to functions.

### ISSUE THREAD

**Issue by davidben** [openssl/openssl#1319]
There's a lot of API which is secretly a bunch of macros over `SSL_CTX_ctrl` or `SSL_ctrl`. Some time ago, we did a pass in BoringSSL and replaced them all with straightforward functions. We actually got rid of the `SSL_CTX_ctrl` and `SSL_ctrl` entry-points altogether. With SSLv2 gone, it doesn't really do much good.

It seems SSL_CTX_ctrl is even documented to not be public API:

>  These functions should never be called directly. All functionalities needed are made available via other functions or macros.

https://www.openssl.org/docs/manmaster/ssl/SSL_CTX_ctrl.html

Benefits of promoting to proper functions:

* Callers are actually type-checked. The current macros aren't type-checked. There's quite a few mistakes that go unnoticed. const-correctness usually goes out the window.
* No more `SSL_CTX_ctrl` in public API.
* More idiomatic API in general. The header file describes the type.
* Adding a new API involves less effort (no adding ctrl value, then entries in both switch-cases and deciding if this is `s3`-only or universal even though they're the same.

Note: there exists code that does things like:
```
#ifdef SSL_CTRL_SET_TLSEXT_HOSTNAME
  SSL_set_tlsext_host_name(ssl, hostname);
#endif
```
or:
```
#ifdef SSL_set_tlsext_host_name
  SSL_set_tlsext_host_name(ssl, hostname);
#endif
```
or:
```
#ifdef SSL_CTRL_SET_TLSEXT_HOSTNAME
  SSL_ctrl(ssl, SSL_CTRL_SET_TLSEXT_HOSTNAME, TLSEXT_NAMETYPE_host_name, hostname);
#endif
```

It's not that common, but I was worried about silent miscompiles, so we added a "Pre-processor compatibility section" to our header which contains a bunch of lines like:

```
#define SSL_CTRL_GET_SESS_CACHE_MODE doesnt_exist
#define SSL_CTRL_GET_SESS_CACHE_SIZE doesnt_exist
#define SSL_CTRL_GET_TLSEXT_TICKET_KEYS doesnt_exist
```
and
```
#define SSL_CTX_add0_chain_cert SSL_CTX_add0_chain_cert
#define SSL_CTX_add1_chain_cert SSL_CTX_add1_chain_cert
#define SSL_CTX_add_extra_chain_cert SSL_CTX_add_extra_chain_cert
```

The other option is to say you're okay with breaking those.

Note that the `doesnt_exist` define will still break the third example (intentionally since we got rid of `SSL_CTX_ctrl`), but the first two will keep working. Alternatively, you could even keep a compatibility `SSL_CTX_ctrl` around which does a switch-case and calls the now real functions (so invert everything), but we haven't found it to be necessary.

**Comment by richsalz** [comment]
Note that for 1.1 we removed all mention of the CTRL macros and documented the public API.  I *think* this is too late for 1.1, but would support it if another team member took it up.

**Comment by mattcaswell** [comment]
@davidben, I agree entirely. This was actually already on my radar for post 1.1.0. I do think it is too late for 1.1.0 though.

**Comment by levitte** [comment]
This cannot be done before 1.2.0, methinks.

**Comment by levitte** [comment]
Having had to deal with diverse macros and functions that are just wrappers around and had to dig through code to see what was actually happening, I've found myself to not be a fan of those wrappers.

So suffice to say that I thoroughly dislike the idea of hiding `SSL_ctrl` or other `_ctrl` functions; I'd rather keep them in the open for those who prefer the direct approach.
Wrappers are still fine for those who do prefer them, I see no reason for stopping that.

Note that this issue hasn't been commented on for two years.  Maybe it's falling off the priority list, maybe not.  I guess we'll see.

I'm labelling it "candidate for closing" for now.  If someone is interesting in driving this issue, feel free to unlabel it.

**Comment by paulidale** [comment]
I like the idea of making them functions (type checking mostly but it is more flexible).  Two confounding issues:

1. Making them inline functions runs into needing also declare them weak for some targets.
2. They are current macros and wouldn't have to stay that way which introduces an extra level of indirection: `#define FOO(a,b,c) foo(a,b,c)` and `static ossl_inline foo(a,b,c) { ... }`

I don't see either as being a big problem.

**Comment by davidben** [comment]
The SSL_ctrl ones are especially goofy because there aren't actually multiple implementations of a given hook. It's a remnant of OpenSSL's SSLv2 heritage and otherwise just serves to make things more complicated and less safe.

One the safety front, calling ctrl functions is not type-specific, making an already difficult API even harder to safely use. The macros aren't type-safe either, but they can be made safe by promoting to proper functions.

In my experience working with an implementation of OpenSSL's API, code that calls into the API, and supporting developers writing more of that code, it's been very clear that the ctrl functions are not a good pattern.

**Comment by levitte** [comment]
Whether SSL_ctrl is to unsafe for public consumption can be argued to be a matter of carefulness and a need for flexibility.  Basically, there's a continuing need to have a direct channel of communication between the application and diverse back ends, and `ctrl` functions are our recipe, and the middle man (libcrypto or libssl) don't necessarily know what it's all about and shouldn't necessarily know.  I personally look at those functions in the same vein as `ioctl`...

Now, I'm not going to defend `SSL_ctrl` specifically, that's a part of the code I'm not familiar enough with to make a stand, so my comments on `ctrl` functions are to be seen from a generic standpoint.  It's entirely possible that `SSL_ctrl` specifically has lost its reason to be with the cancellation of SSLv2, and that all that remains is a role as setter and getter that's independent of the protocol version.  If that is the case, then mark `SSL_ctrl` as deprecated and stop using it, and make all the replacement setter/getter functions that are needed.

Just for clarification, I have nothing against the wrappers themselves, and I have absolutely nothing against making them functions instead of macros.  I do have something against hiding the `ctrl` function itself.

**Comment by davidben** [comment]
The bug didn't propose to hide the non-SSL ctrl functions. Those are also a bad API design, but a much more complex problem because the corresponding method tables are exposed. It only proposed to do away with `SSL_ctrl` which truly has no purpose in life.

**Comment by levitte** [comment]
Unfortunately, you're not quite right, and here are the exhibits:

https://github.com/openssl/openssl/blob/master/ssl/d1_lib.c#L206-L240

https://github.com/openssl/openssl/blob/master/ssl/s3_lib.c#L3390-L3733

One might argue that since s3_lib.c covers all SSLv3 as well as all TLS version since then, there isn't really a need for the "backend" line of thought.  Unfortunately, DTLS comes poking...

**Comment by davidben** [comment]
Right, I dealt with this for BoringSSL long long ago. I will very rarely suggest anything to you all that we didn't already prove viable in BoringSSL. :-P This particular one we proved viable back in 2015.

`ssl3_ctrl` is not just SSL3. There is no `tls1_ctrl`. There is a `tls1_ctx_ctrl`, but it's a `#define` for `ssl3_ctx_ctrl`. There is DTLS, yes, but note that DTLS only ever implements additional controls over SSL3/TLS. You may as well split them out into distinct functions with an `SSL_is_dtls` check on top. Not only does this produce a safer API, but it produces a lighter-weight OpenSSL. The static linker can drop those functions when unused, whereas a giant ctrl function forces it all in.

But even that's not the deciding factor. Yes, there are cases when DTLS and SSL3/TLS differ, namely the "transport" interface that the handshake calls into. That's why you have an `SSL_METHOD`. But `SSL_METHOD` is not (and must not be) exposed externally. You have the freedom to revise it. When you need to add a new split between the two protocols, you can just add a new function pointer to `SSL_METHOD`, retain type-safety.

The only reason to have a `FOO_ctrl` function is if `FOO_METHOD` is publicly exposed. (Even then, I think one can come up with a better design---I dislike this habit of reimplementing ad-hoc dynamic languages on top of C---but it's a more complex problem.)

**Comment by kaduk** [comment]
> can be argued to be a matter of carefulness and a need for flexibility

I am extremely leery of arguments that require carefulness from the application author, especially so in security code.  Modern compilers have some really useful warning detection, and in my opinion it's well worth being able to take advantage of them.
Flexibility can be a counterargument, yes, but I think that David's points about SSL_ctrl specifically seem to neuter the flexibility argument in this case.

**Comment by t8m** [comment]
Unfortunately this is not going to happen in 3.0, it is too late for that.

**Comment by nhorman** [comment]
we've pushed this 4 times now, seems like theres no interest in implementing it.

Marking as inactive, to be closed at the end of 3.4 dev barring further input

**Comment by davidben** [comment]
See #27422 and #27418 for examples of problems that this kind of OpenSSL API anti-pattern causes.

Looking at references to this commit, I see one of OpenSSL's downstream applications also suffered because of this bug not being fixed:
https://github.com/yhirose/cpp-httplib/commit/55e99c40300757fcf0a9f5fdc1ffebaba18db9ca#commitcomment-144754325

When I filed this back in 2016, BoringSSL had already made the suggested change. It has worked very well for us and is extremely backwards-compatible. It can happen incrementally and does not need to get scheduled for a major release. It's even ABI-compatilble; if you keep the `SSL_ctrl` symbol around, existing binaries will work just fine. You can simply start replacing the macros with real functions and introducing the real symbols gradually.

**Comment by t8m** [comment]
Given 4.0 is just 2 releases away I would propose to think about this change for 4.0. Although @davidben is correct that this is backwards compatible change it still feels like something for 4.0.

**Comment by davidben** [comment]
While I'm not going to try to run your project for you, I will observe that this backwards-compatible, incremental solution, to a problem that directly impacts OpenSSL-based applications, and has been punted to the next major release several times now since 2016. I think this has been true of quite a few of the suggestions I have made here.

I might suggest that instead you all look at backwards-compatible, incremental solutions as easy opportunities to start chipping away at the problem incrementally. Otherwise, it seems you'll likely repeat this process: 4.0 will come down the pipe, trying to do this all at once seems daunting, it doesn't make it, so we punt to 5.0 and then 6.0, and so on.

More effective I think would be to simply start fixing functions one at a time, perhaps prioritizing the ones that impact applications. Having done this almost a decade ago with a different descendant of the original libssl, I can tell you that it takes probably minutes to convert any given function. The only hard part is realizing this makes sense to do, which I've already done for you.

On my end, while I am very interested in helping this project out, it takes me a lot of time to write up analyses like this one, #6640, https://github.com/openssl/openssl/issues/15199#issuecomment-1631038583, #5158, and many others. When I see that the project is not interested in acting on this feedback, that's fine, but then it would be more effective for me to focus my rather limited time elsewhere.

**Comment by paulidale** [comment]
I think this is a great suggestion and it should really have been acted upon in the almost **nine years** since it was raised.

The only thing that would confound is code that does this kind of thing:
```c
#ifdef SSL_set_tlsext_host_name
    …
#endif
```

I don't see this as a problem given the obvious workaround:
```c
#ifdef SSL_CTRL_SET_TLSEXT_HOSTNAME
    …
#endif
```

Still, the project is extremely conservative about any potential break for any of the unknown morass of users.

**Comment by davidben** [comment]
> `#ifdef SSL_set_tlsext_host_name`

I mentioned this in the bug I filed originally. We were able to accommodate this in BoringSSL by writing `#define SSL_set_tlsext_host_name SSL_set_tlsext_host_name` for everything that was converted. Not amazing, but it does the job.
